Source/Scene/ModelExperimental/CustomShader.js

```diff
@@ -0,0 +1,298 @@
+  var resource;
+  if (typeof textureValue === "string") {
+    resource = Resource.createIfNeeded(textureValue);
+  } else if (textureValue instanceof Resource) {
+    resource = textureValue;
+  }
+
+  resource
+    .fetchImage()
+    .then(function (image) {
+      customShader._loadedImages.push({
+        uniformName: uniformName,
+        image: image,
+      });
+    })
+    .otherwise(function () {
+      var texture = customShader._textures[uniformName];
+      if (defined(texture) && texture !== customShader._defaultTexture) {
+        texture.destroy();
+      }
+
+      customShader._textures[uniformName] = customShader._defaultTexture;
+    });
+}
+
+function getVariables(shaderText, regex, outputSet) {
+  var match;
+  while ((match = regex.exec(shaderText)) !== null) {
+    var variableName = match[1];
+
+    // Using a dictionary like a set. The value doesn't
+    // matter, as this will only be used for queries such as
+    // if (variableName in set) { ... }
+    outputSet[variableName] = true;
+  }
+}
+
+function findUsedVariables(customShader) {
+  var attributeRegex = /[vf]sInput\.attributes\.(\w+)/g;
+  var attributeSet;
+
+  var vertexShaderText = customShader.vertexShaderText;
+  if (defined(vertexShaderText)) {
+    attributeSet = customShader._usedVariablesVertex.attributeSet;
+    getVariables(vertexShaderText, attributeRegex, attributeSet);
+  }
+
+  var fragmentShaderText = customShader.fragmentShaderText;
+  if (defined(fragmentShaderText)) {
+    attributeSet = customShader._usedVariablesFragment.attributeSet;
+    getVariables(fragmentShaderText, attributeRegex, attributeSet);
+
+    var positionRegex = /fsInput\.(position\w+)/g;
+    var positionSet = customShader._usedVariablesFragment.positionSet;
+    getVariables(fragmentShaderText, positionRegex, positionSet);
+
+    var materialRegex = /material\.(\w+)/g;
+    var materialSet = customShader._usedVariablesFragment.materialSet;
+    getVariables(fragmentShaderText, materialRegex, materialSet);
+  }
+}
+
+/**
+ * Update the value of a uniform declared in the shader
+ * @param {String} uniformName The GLSL name of the uniform. This must match one of the uniforms declared in the constructor
+ * @param {Boolean|Number|Cartesian2|Cartesian3|Cartesian4|Matrix2|Matrix3|Matrix4|String|Resource} value The new value of the uniform.
+ */
+CustomShader.prototype.setUniform = function (uniformName, value) {
+  //>>includeStart('debug', pragmas.debug);
+  Check.typeOf.string("uniformName", uniformName);
+  Check.defined("value", value);
+  if (!defined(this.uniforms[uniformName])) {
+    throw new DeveloperError(
+      "Uniform " +
+        uniformName +
+        " must be declared in the CustomShader constructor."
+    );
+  }
+  //>>includeEnd('debug');
+  var uniform = this.uniforms[uniformName];
+  if (uniform.type === UniformType.SAMPLER_2D) {
+    // Textures are fetched asynchronously and updated in update();
+    fetchTexture2D(this, uniformName, uniform.value);
```

Comment:
@lilleyse I can load a texture the first time, but even though `fetchTexture2D()` is correctly updating the uniformMap, I don't see a visual change in the texture. Is there another step needed to mark the texture as dirty or something? 

[local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=hVcNc9s2Ev0rGM3dDHVRqI/YuavjZE4nWak8UuQ6sl2n6jgUCYmoQIAFQEtyJ//9HgCSlu20nbEtAVgsdt++/fCAalZk4ViKMKGrqOCmH8dU67ncUEHek0WD7s/T5ceYzdj5+Oph3P3ExnosLo/jwfjteJP/fD04/yGE0O/Jx40V2txm1+lseM2m+87+tvfTdjK/zD499M3sZnr8iWHv5jydDvvd6c0o+/JxzCaD8/wLlE3n6+PpfKPHGU8Tt77dz4bxEfb2s5874ZfhxY9Jnh3R0W2xiz+vxdH1bvK2d5FKer06nq9/H16MevurH+jbnxaNdwuxEAPv3Nkup4plVJiIj2hkCkV1SEW05HQqE8oPz+GyUQV11+8jRe4Z3VKFXUG3pNR37faCRSN264HERSaoWjSa9T1tImXmUIqr0L6SKotETEMht8GjVFxoI7PPaZQ8f2NwcBL8sRCEFIJZLfqEuCUh7TY520VZzimRK6KpMUysSVQq9TLFnYEN9RVC7iNeYF2b16oOoG0+G85OSJQkRBd5LpUheJAYurOI1e9XF8w+h6LS3Ct/OMdeOJrM+vNS77dWZUccGX1gx5/e/tyfXkzOLu96w9YzmxeN1Jj8pN3OeRTTDTOGijCWWbvb+4/9XTTKN+2HfxhODQqlEFkic8Ok0M4l7QHPEHwScKYNHF7+RmNDjCRxGok1JQnNqUgsolKQQtNVwQXyolmqnc6G49Ht3bQ/P7sc9yfkNQmyyIBGoNDrD2UMqoewwdk6tfFpEpNSwkReGFJf2DLOyZKSXNHXucwLjoMEuyat0IfdCvrs3TWfj0iwjDQdSC5ViwjLLa6bZJ4yTfDjrfXBY8KxolaDvSjP+d5u5lKb17mSNt/tmq5WwECXHl6eXUz6g7MnLv6FWx8pMgB2g4C1X85mHWM7TkOYR0tbHyUOXg8UjZJDW5vWF71heU6TsHZuLWXiXHNXk0JBS3UjtKbbqJ58L41srofP4laxxNpmtrL2x2nRJALvL/53SSKRkKtPk/G8FLd7ZVbrxztxxGMbO8uzUtBd8kRDgsJ6fjdS0dpFjtgYJpZeFRxhwlYrRM+9V29GPE+jFqH3VO1N6gLFIWPhWANPgGMfq6xwFa32f3K4G8LsyuFLe4BwbC0LmDXZKRTOVk50jhR7BP1legS2MHoxkjldoKC0JE4pzy39oEtL1L8YXvq0gQOIlA/GHCE7IV/9dvymPJuijgbX7uvYpci9dp8tL1SZ2jwogdNoQ11eeMMLkbj0sc/b3Vc7kjAF252HT9zzKlZcRoZso3tbqTthl/yLgI7Bm7CDb9WD4Q7s97XUFm97T1FQTjyi98qZGHTCTou4P1alF3YF6WuFfM19V9YsfeBAGfi2zqmlkHLMi7QuMjDEOFw5WkykHA8E2ab484SsNkw1RVfgmO1nL7GWLKlPHdqjcuHxXlV4MyGxjB+yO4fYtDK6sv4wBGMSxTFLXP/ke4hYtiP0nqwmBb6KZkwkmmSuV60k495vKWothUAtz4CkUb4OxJFKfMEqEEnUsZiZPToXizdUaZs3jl+oA0lRlq2STMtI7RHN0pkwMkaxZWFQIOxJDF2KxXexlHigDGedbXUE3tcBPWo+F1KyWKe2HTjO9N7Vbx8R2+fsEOFLUm8Y+NbXckaFu/0LXVXOv3dXQ7VeHpBmIb75aQHzgps6Qg3zaWhL91WeQEeIfn0GSpgJ+pitwcEKSFpKBmWMDoeMEFWo7LaYXzyjF40WCV7MKCB8PSQ0SZt0O52ONeXbwYzDHqj19Zfe8dsWOe72WpDqHf36zh8j4HRnAcIaz44FHAbnX9iH2LvaRo5RI2MJmrRIjiCjk1hsqGvfwMbKVjoD/+UV6TbJP70hIadibVIHXmUdJN3ZL078V3eGF7TkNORyDQj0Fu3QdXnQ3c8UfozzSgplJ8KvdvDQ35k8/vGHVf+t+vz67u/wtiG2eEOvAxOwlbh6zAzjFFcs+Z6EG+Mpysw9vES4A/vK4azoPt4M5/5y8JiZFwigTThbCWeCkhupeELmVCmkPrmhGDtGMA01xn4gT7thz7UlNCq5LCskbK37Cf5PuKRaFgo0sd29r21kk+D46OjNvzvNVv1yjuJV1eCnU0O1iTBg0o6NVCE5c/G3zvtZCPlQazK2B+kIme4V+INDkE+erFzyoN+UoP4FjCXYVrAUe5Aym8vHg0arcarNntMP1Sz6X5a5yRigBGHYNhTzN7JQt5cFypIJY4yIZYaftg+vnibsnrDk/Xf+ayAxB1o4Qdvkn0GkRePDaRvyL66iV9nSOANcPNpbsbT7YeI3wzA8bWP5/ZtGSo4K9Ezz/wE) -- every 5 seconds the texture should change.